File: 0062-unique-paths/0062-unique-paths.cpp
class Solution {
public:
    
    int countPaths(int currX,int currY,int m,int n,vector<vector<int>>& dp)
    {
        if(currX==m-1 & currY == n-1) return 1;
        else if(currX>=m || currY >=n) return 0;
        else if(dp[currX][currY]!=-1) return dp[currX][currY];
        else return dp[currX][currY]=countPaths(currX+1,currY,m,n,dp) + countPaths(currX,currY+1,m,n,dp);
    }
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(m+1,vector<int>(n+1,-1));
        int num = countPaths(0,0,m,n,dp);
        if(m==1 & n==1) return num;
        return dp[0][0];
    }
};

File: 0062-unique-paths/0062-unique-paths.java
class Solution 
{
    public int uniquePaths(int m, int n) 
    {
        int dp[][] = new int[m][n];
        
        for(int i=0;i<m;i++) dp[i][0] = 1;
        for(int j=0;j<n;j++) dp[0][j] = 1;
        int[][] allDirs = {{-1,0},{0,-1}};
        
        for(int i=1;i<m;i++)
        {
            for(int j=1;j<n;j++)
            {
                for(int ind=0;ind<allDirs.length;ind++)
                if(i+allDirs[ind][0]>=0 && i+allDirs[ind][0] < m && j+allDirs[ind][1] >=0 && j+allDirs[ind][1] < n)
                   dp[i][j] = dp[i][j] + dp[i+allDirs[ind][0]][j+allDirs[ind][1]];
            }
        }
        return dp[m-1][n-1];
    }
}